### PR TITLE
fix(N-04): Arithmetic Edge Cases in Quote-Refresh Path Can Cause Aborts and Stale Orders

### LINKED ISSUE #134: [N-04]: Arithmetic Edge Cases in Quote-Refresh Path Can Cause Aborts and Stale Orders

### PR DESCRIPTION
**Pull request by qalisander**
Addressed, except of sub-issue 1: clamping will introduce complication and new edge cases to be considered, for the sake of exotic price and spread values

Resolves #134

#### PR Checklist

- [x] Tests

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved overflow and precision handling in inventory-based price calculations
  * Enhanced decimal precision management in price computations for edge cases

* **New Features**
  * Added validation to detect and prevent invalid base spread conditions

* **Tests**
  * Expanded test coverage for large inventory scenarios and boundary conditions

[Review Change Stack](https://app.coderabbit.ai/change-stack/OpenZeppelin/openzeppelin-sui-amm/pull/150?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/dapp/contracts/prop-amm/sources/config.move [hunk 1]
```diff
@@ -240,8 +240,7 @@ public(package) fun reservation_mid(
     let base_spread = self.base_spread(mid_price) as u128;
     let skew_bps = self.inventory_skew_bps as u128;
     let imbalance = base_balance_in_quote.diff(quote_balance);
-    let shift =
-        base_spread.checked_mul(imbalance).destroy_or!(abort EPriceOverflow) / total_balance;
+    let shift = base_spread.mul_div(imbalance, total_balance);
     let adjusted_shift = shift * skew_bps / HUNDRED_PERCENT_BPS_U128;
```
Comment (bidzyyys) [inline packages/dapp/contracts/prop-amm/sources/config.move:245]: Also looks like `mul_div`.

File: packages/dapp/contracts/prop-amm/sources/market.move [hunk 2]
```diff
@@ -253,15 +253,16 @@ public(package) fun deepbook_price(
...
     let quote_total = quote_exponent + self.quote.decimals;
     let base_total = base_exponent + self.base.decimals;
     let price = if (quote_total >= base_total) {
-        price.checked_mul(10_u128.pow(quote_total - base_total)).destroy_or!(abort EPriceOverflow)
+        let decimal_adj = 10_u128.pow(quote_total - base_total);
+        decimal_adj.mul_div(base_mantissa * constants::float_scaling_u128(), quote_mantissa)
     } else {
-        price / 10_u128.pow(base_total - quote_total)
+        let decimal_adj = 10_u128.pow(base_total - quote_total);
+        base_mantissa * constants::float_scaling_u128() / quote_mantissa / decimal_adj
```
Comment (bidzyyys) [inline packages/dapp/contracts/prop-amm/sources/market.move:266]: Why not `mul_div`?

**Review by bidzyyys — APPROVED** [review]
Looks good, left some nits.

File: packages/dapp/contracts/prop-amm/sources/config.move (hunk 1 above)
Comment (qalisander) [reply packages/dapp/contracts/prop-amm/sources/config.move:245]: mul_div will upcast to u256. Added comment clarifying why multiplication without upcast is safe

File: packages/dapp/contracts/prop-amm/sources/market.move (hunk 2 above)
Comment (qalisander) [reply packages/dapp/contracts/prop-amm/sources/market.move:266]: mul_div will upcast to u256. Added comment clarifying why multiplication without upcast is safe
